snapcraft/cli/env.py

```diff
@@ -15,41 +15,60 @@
 
+from . import echo
+from snapcraft.internal import errors
 
-class ContainerConfig:
 
-    def __init__(self):
-        """
-        Determines if a container should be used and which remote to choose
+class BuilderEnvironmentConfig:
+    """Handle the chosen build provider.
 
-        Checks the environment variable SNAPCRAFT_CONTAINER_BUILDS:
-          1. SNAPCRAFT_CONTAINER_BUILDS=1 enables local containers
-          2. SNAPCRAFT_CONTAINER_BUILDS=foobar uses the value as a remote
-          3. SNAPCRAFT_CONTAINER_BUILDS=0 or unset, no container is used
-        """
+    To determine the build environment, SNAPCRAFT_BUILD_ENVIRONMENT is
+    retrieved from the environment and used to determine the build
+    provider. If it is not set, a value of `host` is assumed.
 
-        container_builds = os.environ.get('SNAPCRAFT_CONTAINER_BUILDS', '0')
-        # Default remote if it's a truthy value - otherwise it's a remote name
-        try:
-            self._use_container = util.strtobool(container_builds)
-            self._remote = None
-        except ValueError:
-            self._use_container = True
-            # Verbatim name of a remote
-            if not lxd._remote_is_valid(container_builds):
-                raise errors.InvalidContainerRemoteError(container_builds)
-            self._remote = container_builds
+    Valid values are:
 
-    @property
-    def use_container(self):
-        return self._use_container
+    - host: the host will drive the build.
+    - lxd: the host will setup a container to drive the build.
 
-    @property
-    def remote(self):
-        return self._remote
+    Use of the lxd value is equivalent to setting the now deprecated
+    SNAPCRAFT_CONTAINER_BUILDS environment variable to a value that
+    would evaluate to True.
+    Setting this variable to a value that resolves to a non boolean
+    results in an error.
+    """
 
+    def __init__(self) -> None:
+        use_lxd = None
+        container_builds = os.environ.get('SNAPCRAFT_CONTAINER_BUILDS')
+        if container_builds:
+            echo.warning(
+                'The flag SNAPCRAFT_CONTAINER_BUILDS has been deprecated. '
+                'Use SNAPCRAFT_BUILD_ENVIRONMENT=lxd instead.')
+            try:
+                use_lxd = util.strtobool(container_builds)
+            except ValueError:
+                raise errors.SnapcraftEnvironmentError(
+                    'The experimental feature of using non-local LXD remotes '
+                    'with SNAPCRAFT_CONTAINER_BUILDS has been dropped.')
 
-def get_container_config():
-    return ContainerConfig()
+        build_provider = os.environ.get('SNAPCRAFT_BUILD_ENVIRONMENT')
+        if build_provider and use_lxd:
+            raise errors.SnapcraftEnvironmentError(
+                'SNAPCRAFT_BUILD_ENVIRONMENT and SNAPCRAFT_CONTAINER_BUILDS '
+                'cannot be used together.\n'
+                'Unset one of them from the environment and try again.')
```

Comment:
SNAPCRAFT_CONTAINER_BUILDS is deprecated though, perhaps we should point them toward SNAPCRAFT_BUILD_ENVIRONMENT here instead of saying "just pick one." Otherwise they may just be smacked with a deprecation notice if they pick the wrong one :stuck_out_tongue: .

Reply:
I thought about this. The flow is nice though. If I use both I get:
- the deprecation message
- the error about using both

and this only shows up if using both, so there is no picking the wrong one.